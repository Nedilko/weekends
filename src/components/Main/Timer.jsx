import PropTypes from 'prop-types'
import { useContext, useEffect } from 'react'
import { getTimerData } from '@utils/settings'
import Clock from '@components/Clock/Clock'
import SettingsContext from '@store/Settings'
import useCountdown from '@hooks/useCountdown'
import { isFinished } from '@utils/getTimeLeft'

Timer.propTypes = {
  onFinish: PropTypes.func.isRequired,
}

function Timer({ onFinish }) {
  const settings = useContext(SettingsContext)

  const time = useCountdown(getTimerData(settings.data))

  useEffect(() => {
    if (isFinished(time)) {
      onFinish(true)
    }
  }, [time])

  return <Clock time={time} />
}

export default Timer
