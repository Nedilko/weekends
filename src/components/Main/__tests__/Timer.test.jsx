import { render, screen } from '@utils/test-utils'
import { act } from 'react-dom/test-utils'
import Timer from '@components/Main/Timer'
import SettingsContext from '@store/Settings'

jest.mock('@components/Main/ActionText', () => {
  const originalModule = jest.requireActual('@components/Main/ActionText')
  return {
    __esModule: true,
    ...originalModule,
    default: () => {
      return <div data-testid="action-text">Action text</div>
    },
  }
})

jest.mock('@components/Clock/Clock', () => {
  const originalModule = jest.requireActual('@components/Clock/Clock')
  return {
    __esModule: true,
    ...originalModule,
    default: ({ time }) => {
      return <div data-testid="seconds">{time.seconds}</div>
    },
  }
})

describe('Timer', () => {
  const onFinishHandler = jest.fn()
  beforeEach(() => {
    onFinishHandler.mockClear()
    jest.useFakeTimers()
  })
  afterEach(() => {
    jest.useRealTimers()
  })

  it('should render', () => {
    const date = new Date(2022, 4, 23, 18, 12, 11, 0)
    jest.setSystemTime(date)
    const { container } = render(
      <SettingsContext.Provider value={{ data: { day: 5, hour: 18 } }}>
        <Timer onFinish={onFinishHandler} />
      </SettingsContext.Provider>
    )
    expect(screen.getByTestId('action-text')).toBeInTheDocument()
    expect(screen.getByTestId('seconds')).toBeInTheDocument()
    expect(onFinishHandler).not.toBeCalled()
    expect(container).toMatchSnapshot()
  })

  it('should trigger onFinish when time finishes', () => {
    const date = new Date(2022, 4, 23, 18, 0, 0, 0)
    jest.setSystemTime(date)
    render(
      <SettingsContext.Provider value={{ data: { day: 1, hour: 18 } }}>
        <Timer onFinish={onFinishHandler} />
      </SettingsContext.Provider>
    )
    expect(onFinishHandler).toBeCalled()
  })

  it('should trigger onFinish on next tick', () => {
    const date = new Date(2022, 4, 23, 17, 59, 59, 0)
    jest.setSystemTime(date)
    render(
      <SettingsContext.Provider value={{ data: { day: 1, hour: 18 } }}>
        <Timer onFinish={onFinishHandler} />
      </SettingsContext.Provider>
    )
    const secondsLeft = screen.getByTestId('seconds')
    expect(secondsLeft).toHaveTextContent('1')
    act(() => {
      jest.runOnlyPendingTimers()
    })
    expect(secondsLeft).toHaveTextContent('0')
    expect(onFinishHandler).toHaveBeenCalledTimes(1)
  })
})
