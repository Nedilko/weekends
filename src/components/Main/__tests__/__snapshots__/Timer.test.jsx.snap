// Vitest Snapshot v1

exports[`Timer > should render 1`] = `
<div>
  <div
    class="mt-20 flex flex-col items-center"
  >
    <div
      data-testid="action-text"
    >
      Action text
    </div>
    <div
      data-testid="seconds"
    >
      49
    </div>
  </div>
</div>
`;

exports[`Timer should render 1`] = `
<div>
  <div
    class="mt-20 flex flex-col items-center"
  >
    <div
      data-testid="action-text"
    >
      Action text
    </div>
    <div
      data-testid="seconds"
    >
      49
    </div>
  </div>
</div>
`;
